morango/sync/context.py

```diff
@@ -0,0 +1,224 @@
+            transfer_session_id=(
+                self.transfer_session.id if self.transfer_session else None
+            ),
+            filter=str(self.filter),
+            is_push=self.is_push,
+            stage=self.stage,
+            stage_status=self.stage_status,
+            capabilities=self.capabilities,
+        )
+
+    def __setstate__(self, state):
+        """Re-apply dict state after serialization"""
+        sync_session_id = state.get("sync_session_id", None)
+        if sync_session_id is not None:
+            self.sync_session = SyncSession.objects.get(pk=sync_session_id)
+
+        transfer_session_id = state.get("transfer_session_id", None)
+        if transfer_session_id is not None:
+            self.transfer_session = TransferSession.objects.get(pk=transfer_session_id)
+            if self.sync_session is None:
+                self.sync_session = self.transfer_session.sync_session
+
+        sync_filter = state.get("filter", None)
+        if sync_filter is not None:
+            self.filter = Filter(sync_filter)
+
+        self.is_push = state.get("is_push", None)
+        self.stage = state.get("stage", None)
+        self.stage_status = state.get("stage_status", None)
+        self.capabilities = state.get("capabilities", None)
+
+
+class LocalSessionContext(SessionContext):
+    """
+    Class that holds the context for operating on a transfer locally
+    """
+
+    __slots__ = (
+        "request",
+        "is_server",
+    )
+
+    def __init__(self, request=None, **kwargs):
+        """
+        :param request: If acting as the server, it should pass in the request, but the request
+            is not serialized into context. See `is_server` prop for determining if request was
+            passed in.
+        :type request: django.http.request.HttpRequest
+        """
+        capabilities = kwargs.pop("capabilities", [])
+        if request is not None:
+            capabilities = parse_capabilities_from_server_request(request)
+
+        super(LocalSessionContext, self).__init__(capabilities=capabilities, **kwargs)
+        self.request = request
+        self.is_server = request is not None
+
+    def update(self, **kwargs):
+        super(LocalSessionContext, self).update(**kwargs)
+        # this is a "local" session context, so the stage and status should be in sync with the
+        # local transfer session object
+        if "transfer_session" in kwargs:
+            self.stage = self.transfer_session.transfer_stage or self.stage
+            self.stage_status = (
+                self.transfer_session.transfer_stage_status or self.stage_status
+            )
+
+        # finally, since stage and status should be in sync with local transfer session object
+        # we be sure to pass along updates to stage and status. we also `refresh_from_db` too so
+        # context has the up-to-date instance
+        if self.transfer_session:
+            self.transfer_session.refresh_from_db()
```

Comment:
Is it OK that we didn't refresh from the DB before lines 181-185?

Reply:
I think switching `stage` and `stage_status` to properties will help with the discrepancy, because `"transfer_session" in kwargs` should only ever be `True` when the context is first updated with the `TransferSession`. The super's `update` should be validating against that condition. The subtle difference between  `"transfer_session" in kwargs`  and `bool(self.transfer_session)` doesn't really make that clear.

Overall, I didn't love having this `refresh_from_db`, but in my theoretical async operations I knew we needed to make sure the transfer session object is up-to-date.